Core/clim-core/frames.lisp

```diff
@@ -755,46 +756,60 @@ documentation produced by presentations.")
-    ;; Non-standard pane designator fed to the `make-pane'
-    (t `(make-pane ',(first form) :name ',name ,@(cdr form)))))
+(defun %generic-make-or-reinitialize-pane
+    (panes-for-layout abstract-class-name name &rest initargs)
+  (if-let ((pane (alexandria:assoc-value
+                  panes-for-layout name :test #'eq)))
+    (progn
+      (when-let ((parent (sheet-parent pane)))
+        (sheet-disown-child parent pane))
+      (apply #'reinitialize-instance pane initargs))
+    (apply #'make-pane abstract-class-name :name name initargs)))
+
+(defun make-ensure-pane-form (name form panes-for-layout-var)
+  (flet ((generic-make-or-reinitialize-pane ()
+           (if panes-for-layout-var
+               `(%generic-make-or-reinitialize-pane
+                 ,panes-for-layout-var ',(first form) ',name ,@(rest form))
+               `(make-pane ',(first form) :name ',name ,@(rest form)))))
+    (cond ((and (= (length form) 1) ; Single form which is a function call
+                (listp (first form)))
+           `(coerce-pane-name ,(first form) ',name))
+          ((keywordp (first form)) ; Standard pane (i.e `:application')
+           (case (first form)
+             (:application `(make-clim-application-pane
+                             :name ',name ,@(cdr form)))
+             (:interactor `(make-clim-interactor-pane
+                            :name ',name ,@(cdr form)))
+             (:pointer-documentation `(make-clim-pointer-documentation-pane
+                                       :name ',name
+                                       ,@(cdr form)))
+             (:command-menu `(make-clim-command-menu-pane
+                              :name ',name ,@(cdr form)))
+             (otherwise (generic-make-or-reinitialize-pane))))
+          (t ; Non-standard pane designator fed to the `make-pane'
+           (generic-make-or-reinitialize-pane)))))
 
 (defun make-generate-panes-form (class-name menu-bar panes layouts
-                                 pointer-documentation)
+                                 pointer-documentation reinitializep)
   (when pointer-documentation
     (setf panes (append panes
                         '((%pointer-documentation%
                            pointer-documentation-pane)))))
   `(defmethod generate-panes ((fm frame-manager) (frame ,class-name))
      (let ((*application-frame* frame))
        (with-look-and-feel-realization (fm frame)
-         (unless (frame-panes-for-layout frame)
+         ;; Make (or reinitialize) pane instances and establish
+         ;; lexical variables so layout forms can use them.
+         (let* (,@(when reinitializep
+                    `((old-panes (frame-panes-for-layout frame))))
+                ,@(loop for (name . form) in panes
+                        collect `(,name ,(make-ensure-pane-form
+                                          name form
+                                          (when reinitializep 'old-panes)))))
```

Comment:
`(and reinitializep 'old-panes)`

I think that passing the variable name symbol like this makes the code confusing.

Reply:
- please use gensym here for `old-panes`, otherwise (however unlikely) pane name may be `old-panes`

IMO this let* bindings would be more readable if written in this spirit:

```
,@(if reinitializep
      (with-gensyms (gvar)
        (list* `(,gvar (frame-panes-for-layout frame))
               (loop for (name . form) in panes
                     collect `(,name ,(generate-ensure-pane-form
                                       name form gvar)))))
      (loop for (name . form) in panes
            collect `(,name ,(generate-ensure-pane-form name form nil))))
```

Reply:
This change will be in the second pull request.